Add a method for progress logging multiple alignments

clintval opened the pull request:
Often I am iterating through a BAM by query name and find I'm having to manually loop over all records to increment the progress logger. This new method would make logging all the records in a template more convenient.

tfenne commented:
I don't object to this, but I wonder if a slightly different approach would work for you and be more general:

```python
def record_alignments(
        self,
        alns: Iterable[AlignedSegment],
    ) -> bool:
        """Correctly record all records in an iterable of alignments.
        Args:
            alns: an iterable of zero or more alignments
        Returns:
            true if a message was logged, false otherwise
        """
        logged_message: bool = False
        for rec in alns:
            logged_message = self.record_alignment(rec) or logged_message
        return logged_message
```

That could then be called with `progress.record(template.all_recs())`.

clintval commented:
@tfenne that also works. If I went that approach I might suggest making a small API break since we already have a `record_alignment()` and supporting variadic args:

```diff
diff --git a/fgpyo/util/logging.py b/fgpyo/util/logging.py
--- a/fgpyo/util/logging.py
+++ b/fgpyo/util/logging.py
@@ -150,20 +150,23 @@ class ProgressLogger(AbstractContextManager):

     def record_alignment(
         self,
-        rec: AlignedSegment,
+        *recs: AlignedSegment,
     ) -> bool:
         """Correctly record pysam.AlignedSegments (zero-based coordinates).
```

tfenne commented:
> Let me know which option you like the most and I can fixup.

I'm not sure I like the variadic arg version without changing the name (e.g. to `record_alignment*s*`), which is a bigger breaking change.  And without that, I think most people would miss it.

But I'd be fine to add a `record_alignments` _and_ a `record_template` that just vectors to the former.

clintval commented:
@tfenne I thought about this more and chose to ditch the idea of recording "templates". Simply because I didn't want a user to think the logger was incrementing `1` for every template instead of `n` for all the records _in_ the template. So I've changed the title of the PR and added a new method called `record_alignments()` to help out with collections of alignments including templates.

msto commented on fgpyo/util/logging.py:
```diff
@@ -163,6 +164,23 @@ def record_alignment(
 
+    def record_alignments(
+        self,
+        recs: Iterable[AlignedSegment],
```
I think this is a sensible and helpful addition to the library.

My only suggestion is to type this more strictly as a `Collection`, so the user does not inadvertently pass a generator which would be consumed by the function.

```suggestion
        recs: Collection[AlignedSegment],
```

clintval replied:
@msto I do like that idea. What's awkward is almost everything on template returns an iterator even when it doesn't need to (it's just returning things it already has in memory unaltered). So, if I upgrade to `Collection` I'd have to do this:

```python
progress.record_alignments(recs=list(template.all_recs()))
```

Not the worst, but a little awkward.

clintval replied:
I thought a bit more about this and I left the typing as the most generic thing necessary for the function to work, which is `Iterable`. The function is named "record_alignments" so it should be unsurprising that it will advance/consume/iterate-over whatever you supply it.
